File: lockserver/src/jhpc-message-c/vanillaClient.c
#include <stdlib.h>

#include "messageClient.h"
#include "message.h"

#define VANILLA_SERVICE_PORT    1999
#define VANILLA_SERVICE_MESSAGE 100

int main(int argc, char* argv[]) {
	char* host;
	MessageClient conn;
	Message m, reply;
	int lockId = 25;
	int numberOfLocks = 2;
	int offset0 = 10;
	int length0 = 5;
	int offset1 = 20;
	int length1 = 5;
	int64_t long0 = 12345678901234567LL;
	int64_t long1 = -23456789012345678LL;
	int i = 0;
	
	int offsets1[] = {10, 20};
	int lengths1[] = {5, 5};
	int64_t offsets2[] = { 12345678901234567LL, 23456789012345678LL };
	int64_t lengths2[] = { 1000000000000000LL,  1000000000000000LL };
	
	if (argc >= 2) {
		host = argv[1];
	} else {
		host = "localhost";
	}

	if ((conn = newClient(host, VANILLA_SERVICE_PORT)) == NULL) {
		printf("%s: failed to create message client\n", argv[0]);
		return 1;
	}

	m = newMessage();
	m->type = VANILLA_SERVICE_MESSAGE;
	setStringParam(m, "functionName", "acquire");
	setIntegerParam(m, "lockId", lockId);
	setStringParam(m, "fileName", "xyz.dat");
	setIntegerParam(m, "numberOfLocks", numberOfLocks);

	setIntegerParam(m, "offset0", offset0);
	setIntegerParam(m, "length0", length0);

	setIntegerParam(m, "offset1", offset1);
	setIntegerParam(m, "length1", length1);
	
	setLongParam(m, "long0", long0);
	setLongParam(m, "long1", long1);
	
	printf("finished setting initial params\n");
	fflush(stdout);

	createIntegerArrayParam(m, "offsets1", 2);
	printf("finished createIntegerArrayParam\n");
	for (i = 0; i < 2; ++i) {
		setIntegerArrayParamValue(m, "offsets1", i, offsets1[i]);
	}
	
	createIntegerArrayParam(m, "lengths1", 2);
	for (i = 0; i < 2; ++i) {
		setIntegerArrayParamValue(m, "lengths1", i, lengths1[i]);
	}
	
	createLongArrayParam(m, "offsets2", 2);
	for (i = 0; i < 2; ++i) {
		setLongArrayParamValue(m, "offsets2", i, offsets2[i]);
	}
	
	createLongArrayParam(m, "lengths2", 2);
	for (i = 0; i < 2; ++i) {
		setLongArrayParamValue(m, "lengths2", i, lengths2[i]);
	}
	
	printf("Message:\n");
	printMessage(m);
	printf("functionName:%s\n", getStringParam(m, "functionName"));
	printf("lockId:%d\n", getIntegerParam(m, "lockId"));

	reply = clientCall(conn, m);
	printf("Message instance received from server:\n");
	printMessage(reply);
	printf("numberOfLocksGranted:%d\n", getIntegerParam(reply, "numberOfLocksGranted"));
	printf("lockId:%d\n", getIntegerParam(reply, "lockId"));
	clientDisconnect(conn);
	
	destroyMessage(reply);
	destroyMessage(m);
	return 0;
}

File: lockserver/src/info/jhpc/textbook/message2/VanillaClient.java
/*
 * Copyright (c) 2000, Thomas W. Christopher and George K. Thiruvathukal
 * 
 * Java and High Performance Computing (JHPC) Organzization Tools of Computing
 * LLC
 * 
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * 
 * The names Java and High-Performance Computing (JHPC) Organization, Tools of
 * Computing LLC, and/or the names of its contributors may not be used to
 * endorse or promote products derived from this software without specific prior
 * written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 * This license is based on version 2 of the BSD license. For more information
 * on Open Source licenses, please visit http://opensource.org.
 */

package info.jhpc.textbook.message2;

import java.io.IOException;
import info.jhpc.message2.Message;
import info.jhpc.message2.MessageClient;

public class VanillaClient {

	public static void main(String[] args) {
		String host;
		try {
			host = args[0];
		} catch (ArrayIndexOutOfBoundsException e) {
			host = "localhost";
		}
		int port;
		try {
			port = Integer.parseInt(args[1]);
		} catch (ArrayIndexOutOfBoundsException e) {
			port = VanillaService.VANILLA_SERVICE_PORT;
		}

		MessageClient conn;
		try {
			conn = new MessageClient(host, port);
		} catch (IOException e) {
			System.err.println("Could not contact VanillaService @ " + host
					+ ":" + port + ": " + e.getMessage());
			e.printStackTrace();
			return;
		}

		Message m = new Message();
		m.setType(VanillaService.VANILLA_SERVICE_MESSAGE);
		m.setStringParam("functionName", "acquire");
		m.setIntegerParam("lockId", 25);
		m.setStringParam("fileName", "xyz.dat");
		m.setIntegerParam("numberOfLocks", 2);

		m.setIntegerParam("offset0", 10);
		m.setIntegerParam("length0", 5);

		m.setIntegerParam("offset1", 20);
		m.setIntegerParam("length1", 5);
		
		m.setLongParam("long0", 12345678901234567L);
		m.setLongParam("long1", -23456789012345678L);

		m = conn.call(m);
		System.out.println("Message instance received from server " + m);
		conn.disconnect();
	}
}